services/copier/service.go

```diff
@@ -0,0 +1,261 @@
+	// Read request length.
+	var n uint32
+	if err := binary.Read(r, binary.BigEndian, &n); err != nil {
+		return nil, fmt.Errorf("read request length: %s", err)
+	}
+
+	// Read body.
+	buf := make([]byte, n)
+	if _, err := io.ReadFull(r, buf); err != nil {
+		return nil, fmt.Errorf("read request: %s", err)
+	}
+
+	// Unmarshal request.
+	req := &internal.Request{}
+	if err := proto.Unmarshal(buf, req); err != nil {
+		return nil, fmt.Errorf("unmarshal request: %s", err)
+	}
+
+	return req, nil
+}
+
+// writeResponse marshals and writes a Response to w.
+func (s *Service) writeResponse(w io.Writer, resp *internal.Response) error {
+	// Marshal the response to a byte slice.
+	buf, err := proto.Marshal(resp)
+	if err != nil {
+		return fmt.Errorf("marshal error: %s", err)
+	}
+
+	// Write response length to writer.
+	if err := binary.Write(w, binary.BigEndian, uint32(len(buf))); err != nil {
+		return fmt.Errorf("write response length error: %s", err)
+	}
+
+	// Write body to writer.
+	if _, err := w.Write(buf); err != nil {
+		return fmt.Errorf("write body error: %s", err)
+	}
+
+	return nil
+}
+
+// Client represents a client for connecting remotely to a copier service.
+type Client struct {
+	host string
+}
+
+// NewClient return a new instance of Client.
+func NewClient(host string) *Client {
+	return &Client{
+		host: host,
+	}
+}
+
+// ShardReader returns a reader for streaming shard data.
+// Returned ReadCloser must be closed by the caller.
+func (c *Client) ShardReader(id uint64) (io.ReadCloser, error) {
+	// Connect to remote server.
+	conn, err := tcp.Dial("tcp", c.host, MuxHeader)
+	if err != nil {
+		return nil, err
+	}
+
+	// Send request to server.
+	if err := c.writeRequest(conn, &internal.Request{ShardID: proto.Uint64(id)}); err != nil {
+		return nil, fmt.Errorf("write request: %s", err)
+	}
+
+	// Read response from the server.
+	resp, err := c.readResponse(conn)
+	if err != nil {
+		return nil, fmt.Errorf("read response: %s", err)
+	}
+
+	// If there was an error then return it and close connection.
+	if resp.GetError() != "" {
+		conn.Close()
+		return nil, errors.New(resp.GetError())
+	}
+
+	// Returning remaining stream for caller to consume.
+	return conn, nil
+}
+
+// writeRequest marshals and writes req to w.
+func (c *Client) writeRequest(w io.Writer, req *internal.Request) error {
+	// Marshal request.
+	buf, err := proto.Marshal(req)
+	if err != nil {
+		return fmt.Errorf("marshal request: %s", err)
+	}
+
+	// Write request length.
+	if err := binary.Write(w, binary.BigEndian, uint32(len(buf))); err != nil {
+		return fmt.Errorf("write request length: %s", err)
+	}
+
+	// Send request to server.
+	if _, err := w.Write(buf); err != nil {
+		return fmt.Errorf("write request body: %s", err)
+	}
+
+	return nil
+}
+
+// readResponse reads and unmarshals a Response from r.
+func (c *Client) readResponse(r io.Reader) (*internal.Response, error) {
+	// Read response length.
+	var n uint32
```

Comment:
That's just for the response header. The shard writes out its data size as a `uint64`.